### PR TITLE
Plot window api

### PR DESCRIPTION
**Pull request by mortalisk**
While working on #602, the plot window code has been cleaned up somewhat.

The PlotApi class has been introduced, as a temporary api for the plot window (and potentially more of the gui) to consume in a semi-server-client-style-stateless-kind-of-fashion.

The point of this pull request is not the PlotApi as such, but to maybe preserve the improvements to the readability of the plot code that have been made, which clarifies the requirements of the plot code, and the data access patterns it uses.

### CODE REVIEW

File: ert_data/obs_loader.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    elif observation_type == "SUMMARY_OBS":
+        return load_summary_data
+    elif observation_type == "BLOCK_OBS":
+        return load_block_data
+    else:
+        raise TypeError("Unknown observation type: {}".format(observation_type))
+
+
+def load_general_data(facade, observation_key, case_name):
+    obs_vector = facade.get_observations()[observation_key]
```
Comment (xjules) [inline ert_data/obs_loader.py]: if `observation_key` in `facade.get_observations()` ?

File: ert_data/obs_loader.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        raise TypeError("Unknown observation type: {}".format(observation_type))
+
+
+def load_general_data(facade, observation_key, case_name):
+    observations = facade.get_observations()
+    data = pd.DataFrame()
+    if observation_key not in observations:
+        return data
+
+    obs_vector = facade.get_observations()[observation_key]
```
Comment (xjules) [inline ert_data/obs_loader.py]: `facade.get_observations()` should become now `observations`

Comment (mortalisk) [reply ert_data/obs_loader.py]: Of course. That was a test. You passed :p

Comment (mortalisk) [reply ert_data/obs_loader.py]: I'll also remove that misleading comment a few lines later

File: ert_data/obs_loader.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        # the simulation data.
+        # You then get something like:
+        #      0   1   2
+        # OBS  NaN NaN 42
+        # STD  NaN NaN 4.2
+        #   0  1.2 2.3 7.5
+        node = obs_vector.getNode(time_step)
+        index_list = [node.getIndex(nr) for nr in range(len(node))]
+
+        data = (
```
Comment (xjules) [inline ert_data/obs_loader.py]: maybe there is a reason to write it like this - but why not just put `get_data_points` and `get_std()` as two columns directly?

File: ert_data/obs_loader.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        return load_general_data
+    elif observation_type == "SUMMARY_OBS":
+        return load_summary_data
+    elif observation_type == "BLOCK_OBS":
+        return load_block_data
+    else:
+        raise TypeError("Unknown observation type: {}".format(observation_type))
+
+
+def load_general_data(facade, observation_key, case_name):
```
Comment (xjules) [inline ert_data/obs_loader.py]: Shall we rename it to `load_obs_data`?

File: ert_data/results.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+from ert_data import loader
+from ert_data.measured import MeasuredData, _add_index_range
+
+
+class Results:
+
+    def __init__(self, facade, keys, case):
+        self._facade = facade
+        self._case = case
+        self._set_data(self._get_data(keys))
```
Comment (xjules) [inline ert_data/results.py]: `self.data = self._get_data ...` ?

File: ert_data/results.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    @property
+    def data(self):
+        return self._data
+
+    def _set_data(self, data):
```
Comment (xjules) [inline ert_data/results.py]: Don't understand how `_set_data` and `_get_data` refer to `self._data`

File: ert_data/results.py (hunk 5 above)
Comment (mortalisk) [reply ert_data/results.py]: Just an artifact of copy-pasta oriented programming :p trying to keep the code similar to the existing MeasuredData class. Not sure where we are going with this code to be honest, so not sure how to structure it. Perhaps we could just move the logic to the facade, and simply call it from the PlotApi. Let's discuss how it should be.

File: ert_data/results.py (hunk 6 above)
Comment (mortalisk) [reply ert_data/results.py]: They are called in the constructor. I think it should be changed to simpler code.

File: ert_gui/plottery/plot_config_factory.py [hunk 7]
```diff
@@ -4,34 +4,24 @@
...
-    def updatePlotConfigForKey(cls, ert, key, plot_config):
-        """
-        @type ert: res.enkf.enkf_main.EnKFMain
-        @param key: str
-        @return: PlotConfig
-        """
-        key_manager = ert.getKeyManager()
         # The styling of statistics changes based on the nature of the data
-        if key_manager.isSummaryKey(key) or key_manager.isGenDataKey(key):
+        if key["dimentionality"] == 2:
```
Comment (xjules) [inline ert_gui/plottery/plot_config_factory.py]: typo: dimensionality

File: ert_gui/plottery/plots/gaussian_kde.py [hunk 8]
```diff
@@ -43,21 +46,21 @@ def _plotGaussianKDE(axes, plot_config, data, label):
 
     style = plot_config.histogramStyle()
 
-    if data.dtype == "object":
-        try:
-            data = pd.to_numeric(data, errors='coerce')
-        except AttributeError:
-            data = data.convert_objects(convert_numeric=True)
+    # if data.dtype == "object":
```
Comment (oysteoh) [inline ert_gui/plottery/plots/gaussian_kde.py]: Remove lines not in use anymore

File: ert_data/obs_loader.py (hunk 4 above)
Comment (oysteoh) [reply ert_data/obs_loader.py]: case_name not used

File: ert_data/obs_loader.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        "OBS": data_points,
+        "STD": std
+    }
+
+    df = pd.DataFrame(data=data, index=index_list).T
+
+    return df
+
+
+def load_block_data(facade, observation_key, case_name):
```
Comment (oysteoh) [inline ert_data/obs_loader.py]: case_name not used

File: ert_data/results.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+            data = pd.concat({key: data}, axis=1)
+
+            measured_data = pd.concat([measured_data, data], axis=1)
+
+        try:
+            return measured_data.astype(float)
+        except ValueError:
+            return measured_data
+
+    def _data_for_key(self, key):
```
Comment (oysteoh) [inline ert_data/results.py]: what if all resolves to False?

File: ert_data/results.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        self._data = data
+
+    def _get_data(self, data_keys):
+        measured_data = pd.DataFrame()
+
+        for key in data_keys:
+
+            data = self._data_for_key(key)
+
+            # Simulated data and observations both refer to the data
```
Comment (oysteoh) [inline ert_data/results.py]: remove code not used

File: ert_data/results.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
+import pandas as pd
+
+from ert_data import loader
+from ert_data.measured import MeasuredData, _add_index_range
+
```
Comment (oysteoh) [inline ert_data/results.py]: loader and _add_index_range not in use

File: ert_data/obs_loader.py (hunk 9 above)
Comment (mortalisk) [reply ert_data/obs_loader.py]: The file follows the pattern of the loader.py file. The case is required in one case, and not the others, but is there to keep the interface the same for the three functions.

**Comment by jondequinor** [comment]
An attempt at combining `loader` and `obs_loader` should be made. Maybe one loader can serve both use cases (plot and obs_scaling).

**Comment by jondequinor** [comment]
```
$ ert gui snake_oil.ert 
Traceback (most recent call last):
  File "/data/work/ert/ert_gui/tools/plot/plot_tool.py", line 12, in trigger
    plot_window = PlotWindow(self._config_file, self.parent())
  File "/data/work/ert/ert_gui/tools/plot/plot_window.py", line 39, in __init__
    self._key_definitions = self._api.all_data_type_keys()
  File "/data/work/ert/ert_gui/tools/plot/plot_api.py", line 17, in all_data_type_keys
    for key in self._facade.all_data_type_keys()]
  File "/data/work/ert/ert_shared/libres_facade.py", line 117, in observation_keys
    return SummaryObservationCollector.observationKeys(self._enkf_main, key)
AttributeError: type object 'SummaryObservationCollector' has no attribute 'observationKeys'
```

File: ert_shared/libres_facade.py [hunk 13]
```diff
@@ -81,3 +83,114 @@ def select_or_create_new_case(self, case_name):
...
+
+            obs_key = GenDataObservationCollector.getObservationKeyForDataKey(self._enkf_main, key, report_step)
+            if obs_key is not None:
+                return [obs_key]
+            else:
+                return []
+        elif self._enkf_main.getKeyManager().isSummaryKey(key):
+            return SummaryObservationCollector.observationKeys(self._enkf_main, key)
+        else:
+            return []
```
Comment (oyvindeide) [inline ert_shared/libres_facade.py]: There is quite extensive logic here which should be tested.

File: ert_shared/libres_facade.py [hunk 14]
```diff
@@ -81,3 +83,114 @@ def select_or_create_new_case(self, case_name):
...
+        if refcase is None or key not in refcase:
+            return DataFrame()
+
+        values = refcase.numpy_vector(key, report_only=False)
+        dates = refcase.numpy_dates
+
+        data = DataFrame(zip(dates, values), columns=['Date', key])
+        data.set_index("Date", inplace=True)
+
+        return data.iloc[1:]
```
Comment (oyvindeide) [inline ert_shared/libres_facade.py]: There is quite extensive logic here which should be tested.

File: ert_shared/libres_facade.py [hunk 15]
```diff
@@ -81,3 +83,114 @@ def select_or_create_new_case(self, case_name):
...
+            report_step = int(key_parts[1])
+        else:
+            report_step = 0
+
+        try:
+            data = GenDataCollector.loadGenData(self._enkf_main, case, key, report_step)
+        except ValueError:
+            data = DataFrame()
+
+        return data.dropna() # removes all rows that has a NaN
```
Comment (oyvindeide) [inline ert_shared/libres_facade.py]: There is quite extensive logic here which should be tested.

File: ert_shared/libres_facade.py [hunk 16]
```diff
@@ -81,3 +83,114 @@ def select_or_create_new_case(self, case_name):
...
+
+            if any(data.duplicated()):
+                print("** Warning: The simulation data contains duplicate "
+                      "timestamps. A possible explanation is that your "
+                      "simulation timestep is less than a second.")
+                data = data.drop_duplicates()
+
+            data = data.pivot(index="Date", columns="Realization", values=key)
+
+        return data #.dropna()
```
Comment (oyvindeide) [inline ert_shared/libres_facade.py]: There is quite extensive logic here which should be tested.

File: ert_gui/tools/plot/plot_api.py [hunk 17]
```diff
@@ -0,0 +1,124 @@
+from ert_data import loader as loader
+import pandas as pd
+
+
+class PlotApi:
```
Comment (oyvindeide) [inline ert_gui/tools/plot/plot_api.py]: This class should be tested as it's central to the GUI and potentially the future data api.

File: ert_shared/libres_facade.py (hunk 14 above)
Comment (mortalisk) [reply ert_shared/libres_facade.py]: Added test_libres_facade with tests for the new methods

File: ert_gui/ertwidgets/checklist.py [hunk 18]
```diff
@@ -93,6 +93,7 @@ def modelChanged(self):
         items = self._model.getList()
 
         for item in items:
+            item = item["key"]
```
[the remark this answers is not in the document]
Comment (mortalisk) [reply ert_gui/ertwidgets/checklist.py]: Was sending in the dict from the API to this class, which was not necessary. Removed this change and gave the checklist only the actual keys to work on instead.

File: ert_shared/libres_facade.py [hunk 19]
```diff
@@ -81,3 +83,120 @@ def select_or_create_new_case(self, case_name):
...
+        elif self._enkf_main.getKeyManager().isSummaryKey(key):
+            return [str(k) for k in self._enkf_main.ensembleConfig().getNode(key).getObservationKeys()]
+        else:
+            return []
+
+    def gather_gen_kw_data(self, case, key):
+        """ :rtype: pandas.DataFrame """
+        data = GenKwCollector.loadAllGenKwData(self._enkf_main, case, [key])
+        if key in data:
+            return data[key].dropna()
```
Comment (sondreso) [inline ert_shared/libres_facade.py]: This returns a Pandas.Series not a Pandas.DataFrame .

Comment (mortalisk) [reply ert_shared/libres_facade.py]: You are correct. I have simply copied the code from the old PlotDataGatherer. Should I add the correct annotation or remove it?

Comment (sondreso) [reply ert_shared/libres_facade.py]: I think a dataframe is reasonable. You could just change to `data[key].to_frame().dropna()`

**Pull request merged**